algorithms/src/std_algorithms/Kokkos_ModifyingOperations.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_MODIFYING_OPERATIONS_HPP
+#define KOKKOS_MODIFYING_OPERATIONS_HPP
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_BeginEnd.hpp"
+#include "Kokkos_Constraints.hpp"
+
+namespace Kokkos {
+namespace Experimental {
+
+// move
+template <typename T>
+KOKKOS_INLINE_FUNCTION std::remove_reference_t<T>&& move(T&& t) {
+  return static_cast<std::remove_reference_t<T>&&>(t);
+}
+
+// swap
+template <class T>
+KOKKOS_INLINE_FUNCTION void swap(T& a, T& b) noexcept {
```

Comment:
Note:  this `noexcept` specification differs from the one in the standard.  At a minimum we should document that.

C++20 changes how swap is found.  Users are no longer required to specialized std::swap; rather, they just have to put a non-member function in the same namespace as their type.  Should we support that kind of lookup here?

Reply:
I added that to be consistent with std syntax - i have to read more about this.

Reply:
Because we need operations that cannot fail as building blocks, `noexcept` really matters for the following operations (as far as optimizations made in the standard library):

Move construction
Move assignment
Swap


For `swap`, `noexcept` is conditional based on the `noexcept`-ness of the underlying move operations until C++17, then uses the `is_nothrow_swappable` trait (with the change in how `swap` is called as I described in the note above).

Also, there is a subtle difference between `noexcept` and not throwing.  `noexcept` has semantics: an exception will not escape a function marked `noexcept`; rather, if one is about to escape, `std::terminate` is called.  In the above, I'm specifically talking about functions marked `noexcept`.